boa_temporal/src/date.rs

```diff
@@ -0,0 +1,245 @@
+        let result = self
+            .iso
+            .add_iso_date(&DateDuration::new(0f64, 0f64, 0f64, days), overflow)?;
+
+        Ok(Self::new_unchecked(result, self.calendar().clone()))
+    }
+
+    #[inline]
+    /// Returns a new moved date and the days associated with that adjustment
+    pub(crate) fn move_relative_date(
+        &self,
+        duration: &Duration,
+        context: &mut dyn Any,
+    ) -> TemporalResult<(Self, f64)> {
+        let new_date = self.add_date(duration, ArithmeticOverflow::Constrain, context)?;
+        let days = f64::from(self.days_until(&new_date));
+        Ok((new_date, days))
+    }
+}
+
+// ==== Public API ====
+
+impl Date {
+    /// Creates a new `Date` while checking for validity.
+    pub fn new(
+        year: i32,
+        month: i32,
+        day: i32,
+        calendar: CalendarSlot,
+        overflow: ArithmeticOverflow,
+    ) -> TemporalResult<Self> {
+        let iso = IsoDate::new(year, month, day, overflow)?;
+        Ok(Self::new_unchecked(iso, calendar))
+    }
+
+    #[must_use]
+    /// Creates a `Date` from a `DateTime`.
+    pub fn from_datetime(dt: &DateTime) -> Self {
+        Self {
+            iso: dt.iso_date(),
+            calendar: dt.calendar().clone(),
+        }
+    }
+
+    #[inline]
+    #[must_use]
+    /// Returns this `Date`'s year value.
+    pub const fn year(&self) -> i32 {
+        self.iso.year()
+    }
+
+    #[inline]
+    #[must_use]
+    /// Returns this `Date`'s month value.
+    pub const fn month(&self) -> u8 {
+        self.iso.month()
+    }
+
+    #[inline]
+    #[must_use]
+    /// Returns this `Date`'s day value.
+    pub const fn day(&self) -> u8 {
+        self.iso.day()
+    }
+
+    #[inline]
+    #[must_use]
+    /// Returns the `Date`'s inner `IsoDate` record.
+    pub const fn iso_date(&self) -> IsoDate {
+        self.iso
+    }
+
+    #[inline]
+    #[must_use]
+    /// Returns a reference to this `Date`'s calendar slot.
+    pub fn calendar(&self) -> &CalendarSlot {
+        &self.calendar
+    }
+
+    /// 3.5.7 `IsValidISODate`
+    ///
+    /// Checks if the current date is a valid `ISODate`.
+    #[must_use]
+    pub fn is_valid(&self) -> bool {
+        self.iso.is_valid()
+    }
+
+    /// `DaysUntil`
+    ///
+    /// Calculates the epoch days between two `Date`s
+    #[inline]
+    #[must_use]
+    pub fn days_until(&self, other: &Self) -> i32 {
+        other.iso.to_epoch_days() - self.iso.to_epoch_days()
+    }
+}
+
+impl IsoDateSlots for Date {
+    /// Returns the structs `IsoDate`
+    fn iso_date(&self) -> IsoDate {
+        self.iso
+    }
+}
+
+// ==== Context based API ====
+
+impl Date {
+    /// Returns the date after adding the given duration to date.
+    ///
+    /// Temporal Equivalent: 3.5.13 `AddDate ( calendar, plainDate, duration [ , options [ , dateAdd ] ] )`
+    #[inline]
+    #[cfg(feature = "context")]
```

Comment:
I don't think adding a feature that gates a parameter is the greatest idea. If two separate crates import `boa_temporal` but one of them uses the `context` feature and the other doesn't, it would break the build for the one that doesn't (cargo unifies features). A better idea would be to remove the feature and add a `contextual_add_to_date` method that has the optional parameter, and `add_to_date` shouldn't take any `context` parameter.